torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp

```diff
@@ -1137,65 +1137,105 @@ RegisterOperators reg2({
+      case TypeKind::FloatType:
+      case TypeKind::StringType:
+      case TypeKind::TensorType:
+        continue;
+      case TypeKind::TupleType:
+        if (!isSortableTupleType(ele_type->expect<TupleType>(), why_not)) {
+          return false;
+        }
+        continue;
+      case TypeKind::ClassType:
+        if (!c10::checkObjectSortSchema(
+                ele_type->expect<ClassType>(), why_not)) {
+          return false;
+        }
+        continue;
+      default:
+        why_not << "Contained elements in " << *tuple_type
+                << " are not sortable. Only Int, Bool, Float, String, Tensor, "
+                << "a User Defined Class with __lt__ method defined or Tuples "
+                << "of aforementionted types can be sorted.";
+        return false;
+    }
+  }
+
+  return true;
 }
 
-Function* checkSortSchema(const c10::TypePtr& list_element_type) {
-  std::stringstream error_str;
-  if (auto class_type = list_element_type->cast<ClassType>()) {
-    if (auto method = class_type->findMethod("__lt__")) {
-      const auto& lt_schema = method->getSchema();
-      const auto& schema_args = lt_schema.arguments();
-      bool error =
-          (schema_args.size() != 2 ||
-           !simpleClassTypeArg(schema_args[0], class_type) ||
-           !simpleClassTypeArg(schema_args[1], class_type) ||
-           lt_schema.returns().size() != 1 ||
-           lt_schema.returns()[0].type() != BoolType::get());
-      if (!error) {
-        return method;
-      }
+bool isSortableListOfObjectsOrTuples(
+    c10::List<IValue>& ivalues,
+    std::stringstream& why_not) {
+  if (ivalues.empty()) {
+    return true;
+  }
+
+  auto type = ivalues.get(0).type();
+  // We assume lists have homogenous types, use first element to determine
+  // best sorting methods. If in the future we need to support heterogenous
+  // types inside list, then sorting needs to have runtime sortable checks.
+  const size_t n = ivalues.size();
+  for (size_t i = 0; i < n; ++i) {
+    const IValue& v = ivalues.get(i);
+    auto curr_type = v.type();
+    if (*curr_type != *type) {
+      why_not << "Only values of same type can be compared. "
+              << "Found " << type->repr_str() << " and "
+              << curr_type->repr_str();
+      return false;
     }
-    error_str << "To sort a list of " << class_type->repr_str()
-              << " it must define a "
-              << "__lt__ method with two inputs of type "
-              << class_type->repr_str() << " that "
-              << "returns a bool";
-  } else {
-    error_str << "To sort a list of " << list_element_type->repr_str()
-              << " must be of Tensors, ints, floats, bools, strs or "
-              << "a User Defined Class that defines the __lt__ compare method"
-              << ", got list of " << list_element_type->repr_str() << "\n";
   }
-  throw std::runtime_error(error_str.str());
+
+  if (auto tuple_type = type->cast<TupleType>()) {
```

Comment:
Can we refactor `isSortableTupleType` to `isSortableType` and then just call that here instead of the two separate branches for Class / Tuple ? It would be a very small change - change input of isSortableTupleType to a single type instead of tuple_type and not iterate and it works

Reply:
This is actually intentional with a very subtle reason. 

If we use `isSortableType` which, by definition, returns true for other basic types (numbers, strings and tensors), that creates the hazard of someone mistakenly messed up schema matching of `aten::sort.basic_types` and we end up missing the chance of throwing an error here that clearly says type is incorrect because now we need to return true for them. Not a huge deal, but would like to know what you think on thisl.